### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies
import pandas as pd

# stores filepath in a variable
file = "Resources/purchase_data.csv"

# reads file and stores into pandas data frame
purchase_data = pd.read_csv(file)

#prints first 5 rows
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [2]:
# finds unique players
unique_players = purchase_data['SN'].unique()

# calculates total number of unique players
total_players = len(unique_players)

# creates a summary df for total number of players
total_players_summary_df = pd.DataFrame ({"Total Players" : [total_players]})

# displays the total number of players
total_players_summary_df


Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [3]:
#calculates number of unique items 
unique_items = len(purchase_data['Item Name'].unique())

# calculates average purchase price
average_price = purchase_data['Price'].mean()

# calculates total number of purchases
total_purchases = purchase_data['Purchase ID'].nunique()

# calculates total revenue
total_revenue = purchase_data['Price'].sum()

# creates a summary df for total purchases
total_summary_df = pd.DataFrame ({"Number of Unique Items" : [unique_items],
                                  "Average Price" : average_price,
                                  "Number of Purchases" : total_purchases,
                                  "Total Revenue" : total_revenue })

# formats the summary df
total_summary_df['Average Price'] = total_summary_df['Average Price'].map("${:.2f}".format)
total_summary_df['Total Revenue'] = total_summary_df['Total Revenue'].map("${:,}".format)

# displays the summary df 
total_summary_df


Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [4]:
# deletes the duplicates of players
unique_df = purchase_data.drop_duplicates('SN')

# finds unique male, female and other players
male_unique = unique_df.loc[unique_df['Gender'] == "Male", :]
female_unique = unique_df.loc[unique_df['Gender'] == "Female", :]
other_unique = unique_df.loc[unique_df['Gender'] == "Other / Non-Disclosed", :]

# calculates the percentage of male, female and other players
male_percentage = (len(male_unique)/len(unique_df))*100
female_percentage = (len(female_unique)/len(unique_df))*100
other_percentage = (len(other_unique)/len(unique_df))*100

# creates a summary data frame
gender_summary_df = pd.DataFrame ({ "Total Count" : [len(male_unique), len(female_unique), len(other_unique)],
                                  "Percentage of Players" : [male_percentage, female_percentage, other_percentage ]})

# changes the index of df
gender_summary_df = gender_summary_df.rename(index={0 :'Male',
                                                     1 :'Female',
                                                     2 : 'Other / Non-Disclosed' })

# formats the summary df
gender_summary_df['Percentage of Players'] = gender_summary_df['Percentage of Players'].map("{:.2f}%".format)

# displays the summary df 
gender_summary_df


Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [5]:
# groups by gender
grouped_gender_price = purchase_data.groupby('Gender', as_index=True)

# calculates purchase count
purchase_count = grouped_gender_price['Gender'].count()

# calculates average purchase price
average_purchase_price = grouped_gender_price['Price'].mean()

# calculates total purchase value
total_purchase_value = grouped_gender_price['Price'].sum()

# calculates unique number of players by gender
gender_count = grouped_gender_price['SN'].nunique()
gender_unique = purchase_data['Gender'].unique()

# calculates total purchase per person
avg_total_purchase_person = total_purchase_value/gender_count

# creates a summary data frame
gender_summary2_df = pd.DataFrame ({"Purchase Count" : purchase_count,
                                    "Average Purchase Price" : average_purchase_price,
                                    "Total Purchase Value" : total_purchase_value,
                                    "Avg Total Purchase per Person" : avg_total_purchase_person},
                                   index = gender_unique)

#adds index name. Big thanks to Andy for helping me add index name in my summary df
gender_summary2_df.index.name = "Gender"

# formats the summary df
gender_summary2_df['Average Purchase Price'] = gender_summary2_df['Average Purchase Price'].map("${:.2f}".format)
gender_summary2_df['Total Purchase Value'] = gender_summary2_df['Total Purchase Value'].map("${:,.2f}".format)
gender_summary2_df['Avg Total Purchase per Person'] = gender_summary2_df['Avg Total Purchase per Person'].map("${:.2f}".format)

# sorts the summary df by gender index
gender_summary2_df = gender_summary2_df.sort_values(by = "Gender") 

# displays the summary df 
gender_summary2_df


Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [6]:
# establishes bins for ages
bins = [0, 9.9, 14, 19, 24, 29, 34, 39, 100]
group_names = ['<10', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '40+']

# categorizes the existing players using the age bins
unique_df['Age Groups'] = pd.cut(unique_df['Age'], bins, labels = group_names, include_lowest = True)

# calculates the numbers and percentages by age group
age_group_count = unique_df['Age Groups'].value_counts()
age_count = unique_df['Age'].count()
percentage_group = (age_group_count*100)/age_count

# creates a summary data frame
age_summary_df = pd.DataFrame ({"Total Count" : age_group_count,
                                "Percentage of Players" : percentage_group })

# formats the summary df
age_summary_df['Percentage of Players'] = age_summary_df['Percentage of Players'].map("{:.2f}%".format)

# sorts the summary df by index
age_summary_df = age_summary_df.sort_index()

# displays the summary df 
age_summary_df


/Users/ainashtoleu/opt/anaconda3/envs/PythonData/lib/python3.6/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [7]:
# categorizes players from purchase data using the age bins
purchase_data['Age Groups'] = pd.cut(purchase_data['Age'], bins, labels = group_names, include_lowest = True)

# groups by age group
grouped_age_group = purchase_data.groupby('Age Groups')

# calculates purchase count
purchase_count_age = grouped_age_group['Age'].count()

# calculates average purchase price
average_purchase_price_age = grouped_age_group['Price'].mean()

# calculates total purchase value
total_purchase_value_age = grouped_age_group['Price'].sum()

# calculates average total purchase per person
unique_count = grouped_age_group['SN'].nunique()
avg_total_purchase_person_age = total_purchase_value_age/unique_count

age_group_unique = purchase_data['Age Groups'].unique()


# creates a summary data frame
age_summary2_df = pd.DataFrame ({"Purchase Count" : purchase_count_age,
                                    "Average Purchase Price" : average_purchase_price_age,
                                    "Total Purchase Value" : total_purchase_value_age,
                                    "Avg Total Purchase per Person" : avg_total_purchase_person_age},
                                   index = age_group_unique)

#adds index name in summary df
age_summary2_df.index.name = "Age Ranges"

# formats the summary df
age_summary2_df['Average Purchase Price'] = age_summary2_df['Average Purchase Price'].map("${:.2f}".format)
age_summary2_df['Total Purchase Value'] = age_summary2_df['Total Purchase Value'].map("${:,.2f}".format)
age_summary2_df['Avg Total Purchase per Person'] = age_summary2_df['Avg Total Purchase per Person'].map("${:.2f}".format)

# sorts the summary df by age range index
age_summary2_df = age_summary2_df.sort_values(by = "Age Ranges") 

# displays the summary df 
age_summary2_df


Purchase Count Average Purchase Price Total Purchase Value  \
Age Ranges                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05            $1,114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [8]:
# groups by SN 
sn_group = purchase_data.groupby('SN')

# calculates purchase count
purchase_count_sn = sn_group['SN'].count() 

# calculates average purchase price
average_purchase_price_sn = sn_group["Price"].mean()

# calculates total purchase value
total_purchase_value_sn = sn_group["Price"].sum()

# creates a summary data frame
spenders_summary_df = pd.DataFrame ({"Purchase Count" : purchase_count_sn,
                                    "Average Purchase Price" : average_purchase_price_sn,
                                    "Total Purchase Value" : total_purchase_value_sn})

# sorts df by total purchase value column
spenders_summary_df = spenders_summary_df.sort_values("Total Purchase Value", ascending = False)

# formats the summary df
spenders_summary_df['Average Purchase Price'] = spenders_summary_df['Average Purchase Price'].map("${:.2f}".format)
spenders_summary_df['Total Purchase Value'] = spenders_summary_df['Total Purchase Value'].map("${:.2f}".format)

# displays the summary df 
spenders_summary_df.head()


Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, average item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [9]:
# retrieves the item ID, name and price columns
purchase_data_item = purchase_data [["Item ID", "Item Name", "Price"]]

# groups by item ID and name
item_group = purchase_data_item.groupby(['Item ID', 'Item Name'])

# calculates purchase count
purchase_count_item = item_group['Item ID'].count() 

# calculates average item price
average_price_item = item_group["Price"].mean()

# calculates total purchase value
total_purchase_value_item = item_group["Price"].sum()

# creates a summary data frame
item_summary_df = pd.DataFrame ({"Purchase Count" : purchase_count_item,
                                    "Item Price" : average_price_item,
                                    "Total Purchase Value" : total_purchase_value_item})

# formats the summary df
item_summary_df['Item Price'] = item_summary_df['Item Price'].map("${:.2f}".format)
item_summary_df['Total Purchase Value'] = item_summary_df['Total Purchase Value'].map("${:.2f}".format)

# sorts df by purchase count column
item_summary_df_sorted = item_summary_df.sort_values("Purchase Count", ascending = False)

# displays the preview of the summary df 
item_summary_df_sorted.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [10]:
# creates a summary data frame
item_summary2_df = pd.DataFrame ({"Purchase Count" : purchase_count_item,
                                    "Item Price" : average_price_item,
                                    "Total Purchase Value" : total_purchase_value_item})

# sorts df by total purchase value column
item_summary2_df = item_summary2_df.sort_values("Total Purchase Value", ascending = False)

# formats the summary df
item_summary2_df['Item Price'] = item_summary2_df['Item Price'].map("${:.2f}".format)
item_summary2_df['Total Purchase Value'] = item_summary2_df['Total Purchase Value'].map("${:.2f}".format)

# displays the preview of the summary df 
item_summary2_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

   # Heroes of Pymoli - fantasy game

### Three observable trends based on given data

* Based on data analysis, we calculated that the game had a total of 576 players that made at least one purchase. The purchses were made predominantly by males, which comprise roughly 84% of the player base. A total of \\$1,967.64 were spent on in-game items by males, with each spending an average of \\$4.07.

In [11]:
gender_summary_df

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%

In [12]:
gender_summary2_df


Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

* If we analyze the data by age categories, we can observe that the vast majority of the buyers are in the 20-24 age category, with 258 players and a 45% share of the player base. On average, each player in this age category spends \\$4.32. Kids younger than 10 years of age and adults older than 40 years of age spent the least amount of money in comparison to the other age categories. Total purchases made by these two age groups amount to only 36 transactions, and about 5% of the total sales volume. The age categories of 15-19 years, and 20-24 years contributed to the game sales the most, with roughly 2/3 of both the sales volume and the player base. If I was the marketing manager for this company, the choice would be obvious to sell to 15-24 year olds rather than kids and adults over 40.

In [13]:
age_summary_df

Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

In [14]:
age_summary2_df

Purchase Count Average Purchase Price Total Purchase Value  \
Age Ranges                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05            $1,114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

* By further analyzing the sales data, we determined that the most profitable item in the game is the "Final Critic". It was purchased 13 times and brought in \\$59.99. The second most profitable item in the game is the "Oathbreaker", with sales of 12 units and revenue of \\$50.76. While those two items are both most popular and most profitable, there is a discrepancy starting with the third item in these lists. 

In [15]:
item_summary2_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

In [16]:
item_summary_df_sorted.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77